### ISSUE TITLE
Builtin OpenSSL is outdated

### ISSUE THREAD

**Issue by est31** [godotengine/godot#2780]
The version of the builtin openssl is [1.0.1h](https://github.com/godotengine/godot/blob/1d7e9c4ba52250db3b0b0fd732e44074252045d2/drivers/builtin_openssl2/openssl/opensslv.h#L32), which fortunately isn't affected by heartbleed anymore, but still outdated. The most recent version is `1.0.1p`.

For the longer term, I think that it would be great if openssl could be updated to the most recent version for every stable release of godot.

**Comment by akien-mga** [comment]
Marking it as a bug as we are therefore likely exposed to all the security vulnerabilities announced since 1.0.1h. See e.g.:
http://advisories.mageia.org/MGASA-2015-0022.html
http://advisories.mageia.org/MGASA-2015-0111.html
http://advisories.mageia.org/MGASA-2015-0246.html
http://advisories.mageia.org/MGASA-2015-0274.html

Note that there is also a 1.0.2 branch now, with 1.0.2d being the latest release (July 2015).

Maintaining a built-in dependency such as OpenSSL will always be a pain though if we try to ensure that security vulnerabilities are patched in a timely manner (especially if we need to provide patch releases of past versions to address those issues).

It could be of course that many of those vulnerabilities are not exposed to Godot's API, but that's still bad press after the recent press exposure of things like heartbleed.

**Comment by reduz** [comment]
argh updating openssl and getting it to compile on every single platform is
such a giant pain in the ass

**Comment by reduz** [comment]
let's wait to see how progress go with mbedtls

**Comment by reduz** [comment]
well, seems progress with mbedtls stalled, and will not accept a PR using libcurl so I will have to tackle this myself in 2.1

**Comment by est31** [comment]
> argh updating openssl and getting it to compile on every single platform is such a giant pain in the ass

You can try to get changes you do to the openssl source code merged upstream, that way you don't have to repeat them every time you update.

Also note that the Openssl 1.0.1 line will cease to be supported with security updates on Dec 31 this year, so best to switch to 1.0.2: https://www.openssl.org/policies/releasestrat.html

**Comment by slapin** [comment]
Please use git submodules against https://github.com/openssl/openssl
That will remove some of maintenance pain.

**Comment by slapin** [comment]
But again, I think dynamic solution would be preferred for distros. The
static way should only be used for binaries on site.
Distro maintainers should have a way to build against their system openssl.
This allows forgetting about openssl update problems
for distros. I'd also be very happy having separate binaries for distros
(fedora + everythiing else), as fedora prefers to be very different from
others in more than one aspect.
Also I think that dlopen-based solution should be researched as possible
way to overcome library issues.

**Comment by akien-mga** [comment]
> Distro maintainers should have a way to build against their system openssl. This allows forgetting about openssl update problems for distros. 

That's the case already. If you don't specify ``openssl=builtin``, it will use the system-wide version on Linux systems.

**Comment by xodene** [comment]
Is there any progress for this issue? I have a game in Google Play and I was sent an email that the OpenSSL version that I assume Godot uses is vulnerable to the Logjam attack. 

Google Play has a deadline for APKs with this version of OpenSSL set for July 2016, therefore all current Godot APKs are affected by this issue. Is it possible to compile without OpenSSL as a temporary fix?

**Comment by akien-mga** [comment]
You can compile the templates with ``openssl=no`` as a workaround. But indeed, updating the builtin openssl version is high priority.

**Comment by akien-mga** [comment]
Apparently all Google Play Developers are getting a notification that their Godot games are vulnerable to logjam attacks and CVE-2015-3194. While this is likely not true per se (OpenSSL is vulnerable, but I don't think the very small usage we're making of it exposes the mentioned vulnerabilities, though I haven't checked myself), we need to fix this ASAP to silence those warnings.

**Comment by akien-mga** [comment]
I've had a very quick look at updating our builtin version to 1.0.1s (current version in the 1.0.1 branch that we are on right now), it does not seem to build out of the box as I get this error:
```
In file included from drivers/builtin_openssl2/crypto/armcap.c:8:0:
drivers/builtin_openssl2/crypto/arm_arch.h:35:6: error: #error "unsupported ARM architecture"
 #    error "unsupported ARM architecture"
      ^
```
I also seem to get conflicts between the system-wide headers and the ones in builtin_openssl2 when trying to build with ``openssl=builtin``. So... needs more work :)

I'd be glad if someone feeling confident about it would give it a go ;)

**Comment by beocat** [comment]
Looks like the error comes from this file: [https://github.com/openssl/openssl/blob/OpenSSL_1_0_1-stable/crypto/arm_arch.h]
And happens because something (_ARM_ARCH_?) is'nt defined?

**Comment by reduz** [comment]
check the openssl config file (it's somewhere inside the openssl dir), this
one needs to be kept

**Comment by est31** [comment]
@akien-mga perhaps try to make a diff or something to the official release tarball the currently used  openssl comes from, then you know which modifications happened, and you perhaps have a base on which to get it work again.

**Comment by akien-mga** [comment]
Yeah that's a good idea, I'll have a look at it over the week-end.

**Comment by akien-mga** [comment]
Sadly the git history is not really useful as the openssl customizations have often been done in huge commits together with other stuff, so it's difficult to update the current version in a way that Godot can work with out of the box.

Here's the history for whoever is interested:
- https://github.com/godotengine/godot/commits/master/drivers/builtin_openssl
- https://github.com/godotengine/godot/commits/master/drivers/builtin_openssl2

Basically the 1.0.1h version was imported, with the main header for each crypto component split from ``crypto/<component>/<component>.h`` to ``openssl/<component>.h``. Makefiles were removed, so one would have to ignore them when diffing the two versions. Then some custom changes were made to the bundled ssl code, so they might need to be reapplied once the builtin version has been updated.

**Comment by akien-mga** [comment]
@reduz Is there a practical reason for moving those headers to ``openssl`` instead of leaving them at their upstream location? That makes maintenance much more bothersome.

**Comment by primesoftware** [comment]
@reduz I haven't forgotten about the mbedTLS route. I actually have it working in a clunky fashion, but 2 other important projects sidetracked me from finishing. I finished one of those projects and the other is almost done.

**Comment by beocat** [comment]
@akien-mga Have you got an answer on why the the headers have been moved to openssl folder? I agree that it makes the maintenance harder. It would be nice to know what the thought behind this moving-of-files was.

**Comment by rofl0r** [comment]
yeah i agree that it would be much better using mbedtls (but not in-tree) or maintaining a patchset (for windows, since all linux distros (and probably mac) already ship openssl) against latest openssl release tarball.
so on windows, the build process could fetch the latest *ssl sources, patch them, and build them.
an option would be to use libressl instead of openssl, which reportedly builds for all platform without issues.

**Comment by rofl0r** [comment]
the "fix" is only temporal (until the next CVE) - so it does not really solve the issue.
additionally importing several KLOC of changes in a single commit could as well be a security issue - who can tell if one of the diff blobs is not actually a new backdoor (like this https://github.com/alerj78/lucky7coin/issues/1 ) ?

**Comment by vnen** [comment]
Reopening this as it needs more discussion. @rofl0r made a good point. Also, it's not building on Windows, so that's one more problem.

**Comment by est31** [comment]
I don't know whether this is clear to everybody, but the commit adds lots of windows blobs. This might also be the reason for windows builds failing. They should be removed. Its just weird that this wasn't catched by the gitignore.

I've done a `diff -r` comparison from the drivers/builtin_openssl2 directory as of commit 925aa086a27fc06d7dacc866fed975046652cf23 to an extracted download of the [official release tarball](https://www.openssl.org/source/openssl-1.0.1s.tar.gz): http://pastebin.com/jSiVX7Xq

There are no differing files, only files which exist in one directory, but not in the other.

The object files are obviously wrong, I don't know about the other stuff. But for example why add `mips3.s` to `crypto/bn/asm`?

**Comment by akien-mga** [comment]
What Windows blobs? Are you sure they're not from your own compilation of the windows template? I don't have any such blob:

```
[akien@cauldron godot.git (master)]$ find -name "*windows*o"
./drivers/windows/semaphore_windows.x11.tools.64.o
./drivers/windows/shell_windows.x11.tools.64.o
./drivers/windows/thread_windows.x11.tools.64.o
./drivers/windows/file_access_windows.x11.tools.64.o
./drivers/windows/dir_access_windows.x11.tools.64.o
./drivers/windows/mutex_windows.x11.tools.64.o
```

**Comment by akien-mga** [comment]
Thanks for checking that it's the same contents as the upstream tarball, I trust @mrezai's contributions but double-checking is never a bad idea, especially for sensitive drivers like openssl :)

**Comment by akien-mga** [comment]
> But for example why add `mips3.s` to `crypto/bn/asm`?

No idea, but `gitk` shows that it was added by 67894806 and not modified further, so I guess it's no longer in OpenSSL 1.0.1s and @mrezai forgot to remove it.

**Comment by akien-mga** [comment]
Here's @est31's list of differences with the windows compiled files and Makefiles removed: http://pastebin.com/rYunpGmy

**Comment by est31** [comment]
Ah yeah, may be my own artifact.

**Comment by akien-mga** [comment]
OpenSSL 1.0.2h will be released on May 3rd: https://mta.openssl.org/pipermail/openssl-announce/2016-April/000069.html

@mrezai If you're around to handle the update in the following days, it would be awesome. Then I'll prepare a 2.0.3 release to, among other bug fixes, reenable openssl support.

**Comment by mrezai** [comment]
@akien-mga I'll update :)

**Comment by akien-mga** [comment]
Closing as fixed by #4275, #4329 and #4540.

The discussion on how to get rid of security-critical code in our source tree can continue in #4267.
